File: LeetCode/Medium/Combination Sum/Go/combinationSum.go
package main

import "fmt"

func combinationSum(candidates []int, target int) [][]int {
	res := [][]int{}
	backtrack(candidates, target, 0, &res, []int{})
	return res
}

func backtrack(candidates []int, target, currIndex int, res *[][]int, temp []int) {
	if target <= 0 {
		if target == 0 {
			*res = append(*res, append([]int{}, temp...))
		}
		return
	}
	for i := currIndex; i < len(candidates); i++ {
		if candidates[i] > target {
			continue
		}
		temp = append(temp, candidates[i])
		backtrack(candidates, target-candidates[i], i, res, temp)
		temp = temp[:len(temp)-1]
	}
}

func main() {
	fmt.Println(combinationSum([]int{2, 3, 6, 7}, 7))
}

File: LeetCode/Medium/Combination Sum/Java/combinationSum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        backtrack(candidates, target, 0, 0, new ArrayList<>(), res);
        return res;
    }

    private void backtrack(int[] candidates, int target, int currSum, int ind, List<Integer> set, List<List<Integer>> res) {
        if (currSum >= target) {
            if (currSum == target) res.add(new ArrayList<>(set));
            return;
        }

        for (int i=ind; i<candidates.length; i++) {
            if (currSum + candidates[i] > target) continue;
            set.add(candidates[i]);
            backtrack(candidates, target, currSum+candidates[i], i, set, res);
            set.remove(set.size()-1);
        }
    }
}
